### PR TITLE
Fix REPL multicolumn alignment bug with multi-byte characters.

### LINKED ISSUE #11834: REPL loses multicolumn alignment with non-ascii
Tabulation in REPL doesn't seem to compute the display width of non-ascii characters correctly.

From sample autocomplete of `:load` showing files in two columns:

```
second                                    connect                                     
𝔘𝔫𝔦𝔠𝔬𝔡𝔢                            catty.scala                                 
t2.out                                    door.scala 
```

### PR DESCRIPTION
**Pull request by steven-barnes**
Fixes scala/bug#11834

### CODE REVIEW

File: test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala [hunk 1]
```diff
@@ -11,27 +11,32 @@ case class VTabby(width: Int = 80, isAcross: Boolean = false, marginSize: Int =
...
-    assert(res(0)(0) startsWith "a")
+    assert(res(0)(0) startsWith complexA)
     assert(res(0)(0) endsWith "c")
   }
   @Test def twoliner() = {
     val sut   = Tabby(width = 40)
-    val items = List("a" * 15, "b" * 15, "c" * 15)
+    val items = List(complexA * 15, frakturB * 15, "c" * 15)
     val res   = sut tabulate items
     assert(res.size == 2)
```
Comment (som-snytt) [inline test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala:32]: The fellow who wrote this was lazy and didn't use `assertEquals(2, res.size)`.

File: test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala [hunk 2]
```diff
@@ -11,27 +11,32 @@ case class VTabby(width: Int = 80, isAcross: Boolean = false, marginSize: Int =
 @RunWith(classOf[JUnit4])
 class TabulatorTest {
 
+  // à̧̄ is composed using 3 combining characters
+  val complexA = "a\u0300\u0304\u0327"
+  // 𝔟 is stored as 2 characters, but displayed as 1 grapheme
+  val frakturB = "\uD835\uDD1F"
```
Comment (som-snytt) [inline test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala:17]: The random thing I tried was
```
scala> "a\u1ab0"
res6: String = a᪰
```
I asked my French guy, who insists on always using diacritics while his wife slacks, but he didn't know anything.

File: src/repl-frontend/scala/tools/nsc/interpreter/shell/Tabulators.scala [hunk 3]
```diff
@@ -50,14 +50,30 @@ trait Tabulator {
...
 }
 
 /** Adjust the column width and number of columns to minimize the row count. */
 trait VariColumnTabulator extends Tabulator {
   override protected def printMultiLineColumns(items: Seq[String]): Seq[Seq[String]] = {
     import SimpleMath._
-    val longest  = (items map (_.length)).max
-    val shortest = (items map (_.length)).min
+    val longest  = (items map (graphemeCount)).max
+    val shortest = (items map (graphemeCount)).min
```
Comment (som-snytt) [inline src/repl-frontend/scala/tools/nsc/interpreter/shell/Tabulators.scala]: `val (longest, shortest) = items.map(graphemeCount).pipe(vs => (vs.max, vs.min))`

File: src/repl-frontend/scala/tools/nsc/interpreter/shell/Tabulators.scala [hunk 4]
```diff
@@ -50,14 +50,30 @@ trait Tabulator {
...
+
+  protected def graphemeCount(s: String): Int = {
+    import java.text.BreakIterator
+    val it = BreakIterator.getCharacterInstance
+    it.setText(s)
+    var count = 0
+    while (it.next != BreakIterator.DONE) {
+      count += 1
+    }
+    count
```
Comment (som-snytt) [inline src/repl-frontend/scala/tools/nsc/interpreter/shell/Tabulators.scala]: `Iterator.continually(it.next()).takeWhile(_ != DONE).size`

**Review by som-snytt — APPROVED** [review]
Thanks! I had forgotten about BreakIterator. Did I look at it once because of javadoc?

The only style nit is parens for `it.next()` because side-effecting. (The whole codebase does it that way.)

But, anything goes in repl-frontend.

**Comment by som-snytt** [comment]
I'm going to start spelling my name with `a᪰`.

Also name my vars with squiggles: `var a᪰` so everyone thinks it's a squiggle from the IDE.

**Comment by steven-barnes** [comment]
I will make all these changes. Do I need to squash my branch?

File: test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala (hunk 1 above)
Comment (steven-barnes) [reply test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala:32]: Should I change all the asserts to use assertEquals?

File: test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala (hunk 2 above)
Comment (steven-barnes) [reply test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala:17]: The REPL doesn't seem to display combining characters properly. Printing them from a Scala app does work...

**Comment by som-snytt** [comment]
Thanks, still LGTM. My other notes were more just to save typing, I didn't mean to cause you more typing. As a footnote, `assertEquals(expected, actual)`, but please don't bother to fix that up here!

You may squash if you would like, and `push -f`.

**Comment by steven-barnes** [comment]
OK, I think everything is squished correctly

File: test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala (hunk 2 above)
Comment (som-snytt) [reply test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala:17]: But wait, I did that in REPL, where it worked. It's not displaying correctly in this Firefox on Windows. (I'm actually in the Ubuntu subsystem, where I just ssh'd to my real Ubuntu box and lost the arrow in zsh which I was just getting used to.)

File: test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala (hunk 1 above)
Comment (som-snytt) [reply test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala:32]: There are probably better good-third-issue ways to employ your time and talents.

**Comment by som-snytt** [comment]
Thanks. I might mention that I intend to donate it upstream in Java to JLine, which last I checked doesn't have it.

**Comment by SethTisue** [comment]
thanks @steven-barnes !

File: test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala (hunk 2 above)
Comment (steven-barnes) [reply test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala:17]: I'm running on MacOS, using Safari. Your a with double caret does not display correctly at all. Not in Safari, not when I paste "a\u1ab0" into the REPL, or use println("a\u1ab0") (both from REPL and in a unit test are wrong). So maybe Mac can't handle \u1ab0?

In the REPL, "a\u0300\u0304\u0327" is displayed as "a" with grave accent, followed by 2 combining characters. println("a\u0300\u0304\u0327") does the same thing in the REPL, but is correct from a unit test.

**Comment by steven-barnes** [comment]
Thanks, on to the next bug! Does this also need to go into Dotty / Scala 3? If so, I can do that.

File: test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala (hunk 2 above)
Comment (steven-barnes) [reply test/junit/scala/tools/nsc/interpreter/TabulatorTest.scala:17]: Actually Safari also doesn't display "a\u0300\u0304\u0327" correctly; I don't know if these issues are important enough for a Scala bug...

**Comment by SethTisue** [comment]
> Does this also need to go into Dotty / Scala 3? If so, I can do that.

I'm not sure, I guess give it a try and see! https://dotty.epfl.ch/#getting-started

**Comment by som-snytt** [comment]
Maybe spell "Martin Odersky" everywhere with umlauts like Motley Crue.

**Pull request merged**